src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -131,18 +141,36 @@ def get_packages_to_cover(
   # https://pytest.org/en/latest/goodpractices.html#test-discovery. In addition to a performance
   # optimization, this ensures that any transitive sources, such as a test project file named
   # test_fail.py, do not unintentionally end up being run as tests.
-  source_root_stripped_test_target_sources = await Get[SourceRootStrippedSources](
-    Address, test_target.address
+  # TODO: generalize this once it's we add precise file args to `fmt` and `lint`. Those will likely
+  # have a slightly different implementation, such as _not_ stripping source roots for most linters,
+  # so it's not clear yet what should be generalized.
+  test_files_snapshot = adaptor.sources.snapshot
+  if isinstance(origin, FilesystemResolvedSpec):
+    # NB: we ensure that `precise_files_specified` is a subset of the original target's `sources`.
+    # It's possible when given a glob filesystem spec that the spec will have resolved files not
+    # belonging to this target - those must be filtered out.
+    precise_files_specified = set(test_files_snapshot.files).intersection(origin.resolved_files)
```

Comment:
I don't think that it is strictly necessary to do this intersection? The Target's snapshot can only contain files relevant to the target, so it should be possible to apply the globs captured on the CLI directly, which has the same effect as the intersection.

Reply:
And in fact, I think that the same thing applies in general here? The interesection will only be non-empty if the Target owned _any_ of the files on the CLI. And if it did, they should be included, regardless of how many other targets own them.

So... do you actually need to associate addresses with Origins at all for this usecase...?

Reply:
> The Target's snapshot can only contain files relevant to the target,

Agreed.

> so it should be possible to apply the globs captured on the CLI directly, which has the same effect as the intersection

Generally, yes, but the one foil is `FilesystemIgnoreSpec`. Those are not preserved in the `TargetAdaptorWithOrigin`, where `Origin` is only one single `FilesystemResolvedSpec`. Ignores apply to every single include glob, e.g. `./pants fmt2 '*.java' '*.py' '!test_*'`, so for us to have access to the `FilesystemIgnoreGlob`s passed, we would need to start storing that as a sequence in the `FilesystemResolvedSpec`.

Much simpler IMO is to preserve what literal files were resolved at the time that resolution happens:

https://github.com/pantsbuild/pants/blob/93c3567abeb7d19818767d3d20977e5e9ee15ca3/src/python/pants/engine/legacy/graph.py#L781-L789
 
With this design, a `FilesystemResolvedSpec` has a field `resolved_files` as a simple tuple of file names:
https://github.com/pantsbuild/pants/blob/93c3567abeb7d19818767d3d20977e5e9ee15ca3/src/python/pants/base/specs.py#L326-L331

--

Why can't we simply pass `origin.resolved_files` / why do we have to do this intersection? Globs, as explained in the comment. If the original spec is `*.py`, but the target only owns `test_strutil.py`, we need to ensure that we don't also try telling Pytest to run `test_dirutil.py`, `test_osutil.py`, etc.

Reply:
I don't think I understand this. It feels like if we're not preserving the information from the CLI in a way that allows us to have both the excludes and the includes available here to apply as PathGlobs, then we could change that, rather than relying on getting the resolved files. But I could be misunderstanding.

I understand the reason for the intersection, but with the complete set of globs from the CLI I'm pretty sure you can do the intersection directly between the globs and the target.